// ng2-comment/lib/math-captcha.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var MathCaptchaComponent = (function () {
    function MathCaptchaComponent() {
    }
    MathCaptchaComponent.prototype.ngOnInit = function () {
        this.number1 = this.random_number();
        this.number2 = this.random_number();
        this.operation = '+';
        this.answer = this.number1 + this.number2 + '';
    };
    MathCaptchaComponent.prototype.reset = function () {
        this.user_input = '';
        this.refresh();
    };
    MathCaptchaComponent.prototype.random_number = function () {
        return Math.floor((Math.random() * 10) + 1);
    };
    MathCaptchaComponent.prototype.refresh = function () {
        this.ngOnInit();
    };
    MathCaptchaComponent = __decorate([
        core_1.Component({
            selector: 'math-captcha',
            template: "\n\t\tGive an answer: {{ number1 | in_word }} {{ operation | to_str }} {{ number2 | in_word }} = \n\n\t\t<input \n\t\t\ttype=\"text\" \n\t\t\tname=\"answer\" \n\t\t\tid=\"answer\"\n\t\t\tclass=\"btn btn-default\" \n\t\t\ttitle=\"Prove you are human by answering to this math question\" \n\t\t\tplaceholder=\" ? \" \n\t\t\t[(ngModel)]=\"user_input\" \n\t\t\tpattern=\"{{ answer }}\"\n\t\t\tsize=\"2\" \n\t\t\trequired\n\t\t>\n\t\t<button \n\t\t\t(click)=\"refresh()\" \n\t\t\tclass=\"btn btn-link\"\n\t\t\ttitle=\"Get Another Math Question\"\n\t\t><span \n\t\t\t\tclass=\"glyphicon glyphicon-refresh\" \n\t\t\t\taria-hidden=\"true\"\n\t\t\t></span> Refresh</button>\n\n\t"
        }), 
        __metadata('design:paramtypes', [])
    ], MathCaptchaComponent);
    return MathCaptchaComponent;
}());
exports.MathCaptchaComponent = MathCaptchaComponent;
