// angular2-cool-http/src/cool-http.service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var http_1 = require("@angular/common/http");
var Rx_1 = require("rxjs/Rx");
var http_header_model_1 = require("./http-header.model");
var cookie_store_service_1 = require("./cookie-store.service");
var http_error_model_1 = require("./http-error.model");
var request_options_interface_1 = require("./request-options.interface");
var CoolHttp = /** @class */ (function () {
    function CoolHttp(_http) {
        this._http = _http;
        this._cookieStore = new cookie_store_service_1.CookieStore();
        this._globalHeaders = [];
        this._requestInterceptors = [];
        this._responseInterceptors = [];
        this._customCookieToHeaders = [];
    }
    Object.defineProperty(CoolHttp.prototype, "baseUrl", {
        get: function () {
            return this._baseUrl;
        },
        enumerable: true,
        configurable: true
    });
    CoolHttp.prototype.registerBaseUrl = function (baseUrl) {
        this._baseUrl = baseUrl;
        if (this._baseUrl[this._baseUrl.length - 1] !== '/') {
            this._baseUrl += '/';
        }
    };
    CoolHttp.prototype.deRegisterBaseUrl = function () {
        this._baseUrl = null;
    };
    CoolHttp.prototype.setWithCredentials = function (status) {
        this._withCredentials = status;
    };
    CoolHttp.prototype.registerGlobalHeader = function (header) {
        this.deregisterGlobalHeader(header.key);
        this._globalHeaders.push(header);
    };
    CoolHttp.prototype.deregisterGlobalHeader = function (headerKey) {
        var indexOfHeader = this._globalHeaders.findIndex(function (header) { return header.key === headerKey; });
        if (indexOfHeader === -1) {
            return false;
        }
        this._globalHeaders.splice(indexOfHeader, 1);
        return true;
    };
    CoolHttp.prototype.removeAllRegisteredGlobalHeaders = function () {
        this._globalHeaders.length = 0;
    };
    CoolHttp.prototype.registerRequestInterceptor = function (requestInterceptor) {
        this._requestInterceptors.push(requestInterceptor);
    };
    CoolHttp.prototype.deregisterRequestInterceptor = function (requestInterceptor) {
        var indexOfItem = this._requestInterceptors.indexOf(requestInterceptor);
        if (indexOfItem === -1) {
            return false;
        }
        this._requestInterceptors.splice(indexOfItem, 1);
        return true;
    };
    CoolHttp.prototype.registerResponseInterceptor = function (responseInterceptor) {
        this._responseInterceptors.push(responseInterceptor);
    };
    CoolHttp.prototype.deregisterResponseInterceptor = function (responseInterceptor) {
        var indexOfItem = this._responseInterceptors.indexOf(responseInterceptor);
        if (indexOfItem === -1) {
            return false;
        }
        this._responseInterceptors.splice(indexOfItem, 1);
        return true;
    };
    CoolHttp.prototype.sendCookieValueInCustomHeader = function (cookieName, customHeaderName) {
        this._customCookieToHeaders.push({
            cookieName: cookieName,
            customHeaderName: customHeaderName
        });
    };
    CoolHttp.prototype.getAsync = function (url, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        return __awaiter(this, void 0, void 0, function () {
            var that;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        that = this;
                        return [4 /*yield*/, that._requestCoreAsync(url, 'GET', null, options, function (url, data, modOptions) {
                                return that._http.get(url, modOptions);
                            })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    CoolHttp.prototype.postAsync = function (url, data, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        return __awaiter(this, void 0, void 0, function () {
            var that;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        that = this;
                        return [4 /*yield*/, that._requestCoreAsync(url, 'POST', data, options, function (url, data, modOptions) {
                                return that._http.post(url, data, modOptions);
                            })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    CoolHttp.prototype.putAsync = function (url, data, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        return __awaiter(this, void 0, void 0, function () {
            var that;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        that = this;
                        return [4 /*yield*/, that._requestCoreAsync(url, 'PUT', data, options, function (url, data, modOptions) {
                                return that._http.put(url, data, modOptions);
                            })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    CoolHttp.prototype.deleteAsync = function (url, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        return __awaiter(this, void 0, void 0, function () {
            var that;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        that = this;
                        return [4 /*yield*/, that._requestCoreAsync(url, 'DELETE', null, options, function (url, data, modOptions) {
                                return that._http['delete'](url, modOptions);
                            })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    CoolHttp.prototype.patchAsync = function (url, data, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        return __awaiter(this, void 0, void 0, function () {
            var that;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        that = this;
                        return [4 /*yield*/, that._requestCoreAsync(url, 'PATCH', data, options, function (url, data, modOptions) {
                                return that._http.patch(url, data, modOptions);
                            })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    CoolHttp.prototype.headAsync = function (url, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        return __awaiter(this, void 0, void 0, function () {
            var that;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        that = this;
                        return [4 /*yield*/, that._requestCoreAsync(url, 'HEAD', null, options, function (url, data, modOptions) {
                                return that._http.head(url, modOptions);
                            })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    CoolHttp.prototype._requestCoreAsync = function (url, method, data, options, action) {
        return __awaiter(this, void 0, void 0, function () {
            var clientHeaders, shouldIntercept, response, errorResponse_1, returnValue;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        options.headers = options.headers || new http_1.HttpHeaders();
                        options.observe = 'response';
                        options.responseType = 'text';
                        url = this._convertUrl(url);
                        options.headers = this._appendGlobalHeaders(options.headers);
                        options.headers = this._tryAppendRegisteredCookiesToCustomHeaders(options.headers);
                        this._modifyOptions(options);
                        clientHeaders = this._convertAngularHeadersToHttpClientHeaders(options.headers);
                        return [4 /*yield*/, this._invokeRequestInterceptorsAsync(url, method, data, clientHeaders)];
                    case 1:
                        shouldIntercept = _a.sent();
                        if (shouldIntercept) {
                            return [2 /*return*/];
                        }
                        options.headers = this._updateAngularHeadersFromHttpClientHeaders(clientHeaders, options.headers);
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 4, , 5]);
                        return [4 /*yield*/, action(url, data, options).toPromise()];
                    case 3:
                        response = _a.sent();
                        return [3 /*break*/, 5];
                    case 4:
                        errorResponse_1 = _a.sent();
                        response = errorResponse_1;
                        return [3 /*break*/, 5];
                    case 5: return [4 /*yield*/, this._invokeResponseInterceptorsAsync(response, url, method, data, clientHeaders)];
                    case 6:
                        shouldIntercept = _a.sent();
                        if (shouldIntercept) {
                            return [2 /*return*/];
                        }
                        if (!response.ok) {
                            throw new http_error_model_1.HttpError(method, url, response.status, response.statusText, JSON.stringify(response.body));
                        }
                        try {
                            returnValue = JSON.parse(response.body);
                        }
                        catch (e) {
                            returnValue = { data: response.body };
                        }
                        return [2 /*return*/, returnValue];
                }
            });
        });
    };
    CoolHttp.prototype.getObservable = function (url, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        var that = this;
        return that._requestCoreObservable(url, 'GET', null, options, function (url, data, modOptions) {
            return that._http.get(url, modOptions);
        });
    };
    CoolHttp.prototype.postObservable = function (url, data, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        var that = this;
        return that._requestCoreObservable(url, 'POST', data, options, function (url, data, modOptions) {
            return that._http.post(url, data, modOptions);
        });
    };
    CoolHttp.prototype.putObservable = function (url, data, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        var that = this;
        return that._requestCoreObservable(url, 'PUT', data, options, function (url, data, modOptions) {
            return that._http.put(url, data, modOptions);
        });
    };
    CoolHttp.prototype.deleteObservable = function (url, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        var that = this;
        return that._requestCoreObservable(url, 'DELETE', null, options, function (url, data, modOptions) {
            return that._http['delete'](url, modOptions);
        });
    };
    CoolHttp.prototype.patchObservable = function (url, data, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        var that = this;
        return that._requestCoreObservable(url, 'PATCH', data, options, function (url, data, modOptions) {
            return that._http.patch(url, data, modOptions);
        });
    };
    CoolHttp.prototype.headObservable = function (url, options) {
        if (options === void 0) { options = request_options_interface_1.DEFAULT_REQUEST_OPTIONS; }
        var that = this;
        return that._requestCoreObservable(url, 'HEAD', null, options, function (url, data, modOptions) {
            return that._http.head(url, modOptions);
        });
    };
    CoolHttp.prototype._requestCoreObservable = function (url, method, data, options, action) {
        return Rx_1.Observable.fromPromise(this._requestCoreAsync(url, method, data, options, action));
    };
    CoolHttp.prototype._convertUrl = function (url) {
        var returnUrl = url;
        if (this._baseUrl) {
            returnUrl = this._baseUrl + returnUrl;
        }
        return returnUrl;
    };
    CoolHttp.prototype._modifyOptions = function (options) {
        if (this._withCredentials) {
            options.withCredentials = true;
        }
    };
    CoolHttp.prototype._appendGlobalHeaders = function (headers) {
        var newHeaders = headers;
        for (var _i = 0, _a = this._globalHeaders; _i < _a.length; _i++) {
            var registeredHeader = _a[_i];
            newHeaders = headers.set(registeredHeader.key, registeredHeader.value);
        }
        return newHeaders;
    };
    CoolHttp.prototype._tryAppendRegisteredCookiesToCustomHeaders = function (headers) {
        var newHeaders = headers;
        for (var _i = 0, _a = this._customCookieToHeaders; _i < _a.length; _i++) {
            var cookieToHeader = _a[_i];
            var cookieValue = this._cookieStore.getCookie(cookieToHeader.cookieName);
            if (cookieValue) {
                newHeaders = headers.set(cookieToHeader.customHeaderName, cookieValue);
            }
        }
        return newHeaders;
    };
    CoolHttp.prototype._invokeRequestInterceptorsAsync = function (url, method, data, headers) {
        return __awaiter(this, void 0, void 0, function () {
            var _i, _a, requestInterceptor, shouldIntercept;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _i = 0, _a = this._requestInterceptors;
                        _b.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3 /*break*/, 4];
                        requestInterceptor = _a[_i];
                        return [4 /*yield*/, requestInterceptor.beforeRequestAsync(url, method, data, headers)];
                    case 2:
                        shouldIntercept = _b.sent();
                        if (shouldIntercept) {
                            return [2 /*return*/, true];
                        }
                        _b.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/, false];
                }
            });
        });
    };
    CoolHttp.prototype._invokeResponseInterceptorsAsync = function (response, url, method, data, headers) {
        return __awaiter(this, void 0, void 0, function () {
            var _i, _a, responseInterceptor, shouldIntercept;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _i = 0, _a = this._responseInterceptors;
                        _b.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3 /*break*/, 4];
                        responseInterceptor = _a[_i];
                        return [4 /*yield*/, responseInterceptor.afterResponseAsync(response, url, method, data, headers)];
                    case 2:
                        shouldIntercept = _b.sent();
                        if (shouldIntercept) {
                            return [2 /*return*/, true];
                        }
                        _b.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/, false];
                }
            });
        });
    };
    CoolHttp.prototype._convertAngularHeadersToHttpClientHeaders = function (headers) {
        return headers.keys().map(function (headerKey) {
            var httpClientHeader = new http_header_model_1.HttpHeader();
            httpClientHeader.key = headerKey;
            httpClientHeader.value = headers.get(headerKey);
            return httpClientHeader;
        });
    };
    CoolHttp.prototype._updateAngularHeadersFromHttpClientHeaders = function (httpClientHeaders, headers) {
        var newHeaders = headers;
        for (var _i = 0, httpClientHeaders_1 = httpClientHeaders; _i < httpClientHeaders_1.length; _i++) {
            var clientHeader = httpClientHeaders_1[_i];
            var headerValue = headers.get(clientHeader.key);
            if (headerValue !== clientHeader.value) {
                newHeaders = headers.set(clientHeader.key, clientHeader.value);
            }
        }
        return newHeaders;
    };
    CoolHttp.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    CoolHttp.ctorParameters = function () { return [
        { type: http_1.HttpClient, },
    ]; };
    return CoolHttp;
}());
exports.CoolHttp = CoolHttp;
